annotations/views/rest_views.py

```diff
@@ -476,25 +479,87 @@ def create(self, request, *args, **kwargs):
-
     @list_route()
     def search(self, request, **kwargs):
         q = request.GET.get('search', None)
         if not q:
-            return Response({'results': []})
-        pos = request.GET.get('pos', None)
-        concepts = goat.Concept.search(q=q, pos=pos, limit=50)
-
-        def _relabel(datum):
-            _fields = {
-                'name': 'label',
-                'id': 'alt_id',
-                'identifier': 'uri'
-            }
-
-            return {_fields.get(k, k): v for k, v in datum.iteritems()}
-        return Response({'results': map(_relabel, [c.data for c in concepts])})
+                return Response({'results': []})
+        force = request.GET.get('force')
 
+        if cache.get(q.lower()) and q and force == None: # If force is true we want to run the query from scratch and not use the cached results
+            results = cache.get(q.lower())
+        else:
+            pos = request.GET.get('pos', None)
+            concepts = goat.Concept.search(q=q, pos=pos, limit=50)
+
+            def _relabel(datum):
+                _fields = {
+                    'name': 'label',
+                    'id': 'alt_id',
+                    'identifier': 'uri'
+                }
+
+                return {_fields.get(k, k): v for k, v in datum.iteritems() }
+            results = map(_relabel, [c.data for c in concepts])
+
+            def _parseViaf(self, uri, new_concepts):
+                data = urllib2.urlopen(uri)
+                tree = e.parse(data)
+                root = tree.getroot()
+                namespace = {'foaf': 'http://xmlns.com/foaf/0.1/', 'ns1': 'http://viaf.org/viaf/terms#', 'owl': 'http://www.w3.org/2002/07/owl#', 'rdf': 'http://www.w3.org/1999/02/22-rdf-syntax-ns#', 'void': 'http://rdfs.org/ns/void#'}
+                find = e.XPath("//ns1:text", namespaces = namespace)
+                concept_name = find(root)[0].text
+                concept_description = find(root)[0].text
+
+                dic = {
+                    'label': concept_name,
+                    'desc': concept_description,
+                    'uri': uri,
+                    'auth': "VIAF"}
+                new_concepts.append(dic)
+
+
+            def _parseHps(self, uri, new_concepts):
+                data = urllib2.urlopen(uri)
+                tree = ET.parse(data)
+                root = tree.getroot()
+                namespace = {'hps': 'http://www.digitalhps.org/'} 
+
+                for entry in root.findall('hps:conceptEntry', namespace):
+                    description = entry.find('hps:description', namespace)
+                    name = entry.find('hps:lemma', namespace)
+                    concept_description = description.text
+                    concept_name = name.text
+
+                dic = {
+                    'label': concept_name,
+                    'desc': concept_description,
+                    'uri': uri,
+                    'auth': "Concept Power"}
```

Comment:
are you only adding the last entry that has been returned into dic? what about the other ones?

Reply:
if there is just one, you should access it like there is just one (or none)

Reply:
I am not sure what you mean here. I parse the xml and find the corresponding tags and then add them to the dict.